examples/systemcalls/systemcalls.c

```diff
@@ -49,33 +71,65 @@ bool do_exec(int count, ...)
     // and may be removed
     command[count] = command[count];
 
-/*
- * TODO:
- *   Execute a system command by calling fork, execv(),
- *   and wait instead of system (see LSP page 161).
- *   Use the command[0] as the full path to the command to execute
- *   (first argument to execv), and use the remaining arguments
- *   as second argument to the execv() command.
- *
-*/
+    /*
+     * TODO:
+     *   Execute a system command by calling fork, execv(),
+     *   and wait instead of system (see LSP page 161).
+     *   Use the command[0] as the full path to the command to execute
+     *   (first argument to execv), and use the remaining arguments
+     *   as second argument to the execv() command.
+     *
+     */
+    pid_t pidval = fork(); // to execute a new child process
+    if (pidval < 0)        // if negative means child process not created successfully
+    {
+        perror("fork() unsuccessful");
+        exit(1);
+    }
+    else if (pidval == 0) // on successfull execution returns 0 to child process
+    {
+        int n = execv(command[0], &command[0]); // do the execution using execv() with the given command
+        if (n < 0)                              // negative returns means unsuccessful process
+        {
+            perror("execv() unsuccessful");
+            exit(1);
+        }
+    }
+    else
+    {
+        int child_process;             // process ID returned to the parent process
+        pidval = wait(&child_process); // wait for child process to finish
+        if (WIFEXITED(child_process))  // if child process exited sucessfully
+        {
+            printf("Child process exited with %d status\n", WEXITSTATUS(child_process)); // this checks the exit status
+            if (WEXITSTATUS(child_process) == 1)
+            {
+                return false;
```

Comment:
Missing cleanup in these code paths.
Also see discussion in class about avoiding duplication between these two very similar functions.